// hadouken-js-adapter/src/api/interappbus/channel/provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelProvider = void 0;
const channel_1 = require("./channel");
const providerProtectedMap = new WeakMap();
class ChannelProvider extends channel_1.ChannelBase {
    constructor(providerIdentity, send) {
        super(providerIdentity, send, providerProtectedMap);
        this.connectListener = () => undefined;
        this.disconnectListener = () => undefined;
        this.connections = [];
    }
    dispatch(to, action, payload) {
        // verify if the client is valid
        if (this.connections.some((c) => c.name === to.name && c.uuid === to.uuid)) {
            return providerProtectedMap.get(this).send(to, action, payload);
        }
        return Promise.reject(new Error(`The client you are trying to dispatch to is disconnected from this provider.
                {uuid: ${to.uuid}, name: ${to.name}}`));
    }
    async processAction(action, payload, senderIdentity) {
        const clientConnected = this.connections.some((identity) => {
            return identity.uuid === senderIdentity.uuid && identity.name === senderIdentity.name;
        });
        if (!clientConnected) {
            throw new Error('This action was sent from a client that is not connected to the provider.');
        }
        return super.processAction(action, payload, senderIdentity);
    }
    async processConnection(senderId, payload) {
        this.connections.push(senderId);
        return this.connectListener(senderId, payload);
    }
    publish(action, payload) {
        return this.connections.map((to) => providerProtectedMap.get(this).send(to, action, payload));
    }
    onConnection(listener) {
        this.connectListener = listener;
    }
    onDisconnection(listener) {
        this.disconnectListener = listener;
    }
    async destroy() {
        const protectedObj = providerProtectedMap.get(this);
        const { channelName, channelId } = protectedObj.providerIdentity;
        this.connections = [];
        await protectedObj.sendRaw('destroy-channel', { channelName });
        this.removeChannel(channelId);
    }
}
exports.ChannelProvider = ChannelProvider;
